### DISCUSSION TITLE
Could NixOS be an alternative to MicroOS?

### DISCUSSION THREAD

**Discussion by otavio** [mysticaltech/terraform-hcloud-kube-hetzner#219] [Ideas]
I am looking for alternatives to host our cluster and fund this project; it is terrific! I want to ask if someone evaluated the use of NixOS as host distribution?

**Comment by phaer** [comment]
Pretty sure no one has, and even as a NixOS user myself, I don't really see the cost-benefit analysis too positive for such a project. MicroOS works well enough for the current approach and we can use it pretty much as-is.

While NixOS offers much more advanced customization-options and tooling, but comes with a far higher learning-curve (not sure if anyone else on the team used nix before). 

Whats your motivation behind this? Do you want to install NixOS via terraform (already possible with various projects, sometimes a bit hacky). Or would you like to run k3s on NixOS? There's already a service module upstream. Or is it about the hetzner-specific features like CCM and so on?

**Reply by phaer** [reply]
If you aren't already invested in terraform, it might also be worth a try to add NixOS images to https://github.com/syself/cluster-api-provider-hetzner, because that one already supports different operating system images (from packer images).

**Reply by otavio** [reply]
We are using terraform and NixOS in Hetzner, to manage a few servers (using colmena to apply changes), and it has been a pretty good experience.

We intend to move from the existing managed K8S solution we are currently using and found this project. It seems pretty mature, and NixOS as host seemed like a possible choice, but the learning curve is indeed an issue.

**Reply by phaer** [reply]
Nice to read, I am using a pretty similar setup for some machines, albeit without kubernetes there for now :) 

I'd say a PR to add NixOS support would definitely be welcome, but as most of the team probably isn't familiar with nix and content with MicroOS, support for MicroOS should be kept as well and that might complicate matters with the current code-base.

**Reply by otavio** [reply]
Sure. I'll evaluate the MicroOS setup and try out the project. Maybe if I see the need, I could provide a NixOS in the future. I am still trying to get K8S fully, so I need to better understand it before digging more into the project.

**Comment by mysticaltech** [comment]
Hello @otavio, I agree with @phaer, as much as NixOS can be nice. MicroOS is ideal for the job, it's container optimized, it's transactional in nature, it's has a low attack surface and comes hardened out of the box, it's updated daily, as no major version updates, it just updates all the time infinitely (it's based on opensuse tumbleweed), and handles reboot-need signaling well to let Kubernetes know that a reboot is needed after an upgrade. So it's really ideal!

**Comment by mysticaltech** [comment]
That's not in our plans for now, we are moving to the more stable LeapMicro.
